File: scripts/ocr_region_overlay.py
#!/usr/bin/env python3
import argparse
import json
import sys
import tkinter as tk

HANDLE_SIZE = 8
MIN_SIZE = 20


class RegionOverlay:
    def __init__(self, root: tk.Tk, initial_rect):
        self.root = root
        self.screen_w = root.winfo_screenwidth()
        self.screen_h = root.winfo_screenheight()
        self.canvas = tk.Canvas(root, bg="#000000", highlightthickness=0)
        self.canvas.pack(fill="both", expand=True)
        self.rect = initial_rect
        self.mode = None
        self.anchor_x = 0
        self.anchor_y = 0
        self.base_rect = None
        self.result = "CANCELLED"
        self.help_text = (
            "Arraste para criar a area | arraste dentro para mover | arraste bordas para redimensionar | Enter confirma | F tela inteira | Esc cancela"
        )

        self.canvas.bind("<ButtonPress-1>", self.on_press)
        self.canvas.bind("<B1-Motion>", self.on_drag)
        self.canvas.bind("<ButtonRelease-1>", self.on_release)
        self.root.bind("<Return>", self.on_confirm)
        self.root.bind("<Escape>", self.on_cancel)
        self.root.bind("<f>", self.on_fullscreen)
        self.root.bind("<F>", self.on_fullscreen)
        self.root.protocol("WM_DELETE_WINDOW", self.on_cancel)

        self.draw()

    def clamp(self, value, low, high):
        return max(low, min(high, value))

    def normalize_rect(self, rect):
        x1, y1, x2, y2 = rect
        left = min(x1, x2)
        top = min(y1, y2)
        right = max(x1, x2)
        bottom = max(y1, y2)
        left = self.clamp(left, 0, self.screen_w)
        top = self.clamp(top, 0, self.screen_h)
        right = self.clamp(right, 0, self.screen_w)
        bottom = self.clamp(bottom, 0, self.screen_h)
        return [left, top, right, bottom]

    def hit_test(self, x, y):
        if not self.rect:
            return None
        x1, y1, x2, y2 = self.rect
        if x1 > x2:
            x1, x2 = x2, x1
        if y1 > y2:
            y1, y2 = y2, y1
        near_left = abs(x - x1) <= HANDLE_SIZE
        near_right = abs(x - x2) <= HANDLE_SIZE
        near_top = abs(y - y1) <= HANDLE_SIZE
        near_bottom = abs(y - y2) <= HANDLE_SIZE
        inside_x = x1 <= x <= x2
        inside_y = y1 <= y <= y2

        if near_left and near_top:
            return "nw"
        if near_right and near_top:
            return "ne"
        if near_left and near_bottom:
            return "sw"
        if near_right and near_bottom:
            return "se"
        if near_left and inside_y:
            return "w"
        if near_right and inside_y:
            return "e"
        if near_top and inside_x:
            return "n"
        if near_bottom and inside_x:
            return "s"
        if inside_x and inside_y:
            return "move"
        return None

    def on_press(self, event):
        self.anchor_x = event.x
        self.anchor_y = event.y
        self.base_rect = self.rect[:] if self.rect else None
        hit = self.hit_test(event.x, event.y)
        self.mode = hit if hit else "create"
        if self.mode == "create":
            self.rect = [event.x, event.y, event.x, event.y]
        self.draw()

    def on_drag(self, event):
        x = self.clamp(event.x, 0, self.screen_w)
        y = self.clamp(event.y, 0, self.screen_h)
        if self.mode == "create":
            self.rect[2] = x
            self.rect[3] = y
        elif self.mode == "move" and self.base_rect:
            dx = x - self.anchor_x
            dy = y - self.anchor_y
            x1, y1, x2, y2 = self.base_rect
            w = x2 - x1
            h = y2 - y1
            nx1 = self.clamp(x1 + dx, 0, self.screen_w - w)
            ny1 = self.clamp(y1 + dy, 0, self.screen_h - h)
            self.rect = [nx1, ny1, nx1 + w, ny1 + h]
        elif self.mode and self.base_rect:
            x1, y1, x2, y2 = self.base_rect
            if "w" in self.mode:
                x1 = x
            if "e" in self.mode:
                x2 = x
            if "n" in self.mode:
                y1 = y
            if "s" in self.mode:
                y2 = y
            self.rect = [x1, y1, x2, y2]
        self.rect = self.normalize_rect(self.rect)
        self.enforce_min_size()
        self.draw()

    def enforce_min_size(self):
        if not self.rect:
            return
        x1, y1, x2, y2 = self.rect
        if x2 - x1 < MIN_SIZE:
            x2 = min(self.screen_w, x1 + MIN_SIZE)
            x1 = max(0, x2 - MIN_SIZE)
        if y2 - y1 < MIN_SIZE:
            y2 = min(self.screen_h, y1 + MIN_SIZE)
            y1 = max(0, y2 - MIN_SIZE)
        self.rect = [x1, y1, x2, y2]

    def on_release(self, _event):
        self.mode = None
        if self.rect:
            self.rect = self.normalize_rect(self.rect)
            self.enforce_min_size()
        self.draw()

    def draw(self):
        self.canvas.delete("all")
        self.canvas.create_rectangle(
            0, 0, self.screen_w, self.screen_h, fill="#000000", stipple="gray50", outline=""
        )
        self.canvas.create_text(
            20, 20, anchor="nw", text=self.help_text, fill="#ffffff", font=("Helvetica", 14, "bold")
        )
        if self.rect:
            x1, y1, x2, y2 = self.rect
            self.canvas.create_rectangle(x1, y1, x2, y2, outline="#22d3ee", width=3, fill="")
            self.canvas.create_text(
                x1,
                max(26, y1 - 10),
                anchor="sw",
                text=f"{int(x2 - x1)}x{int(y2 - y1)}",
                fill="#22d3ee",
                font=("Helvetica", 12, "bold"),
            )
            for hx, hy in [(x1, y1), (x2, y1), (x1, y2), (x2, y2)]:
                self.canvas.create_rectangle(
                    hx - HANDLE_SIZE,
                    hy - HANDLE_SIZE,
                    hx + HANDLE_SIZE,
                    hy + HANDLE_SIZE,
                    fill="#22d3ee",
                    outline="",
                )

    def on_fullscreen(self, _event=None):
        self.result = "FULLSCREEN"
        self.root.destroy()

    def on_cancel(self, _event=None):
        self.result = "CANCELLED"
        self.root.destroy()

    def on_confirm(self, _event=None):
        if not self.rect:
            self.result = "CANCELLED"
        else:
            x1, y1, x2, y2 = self.normalize_rect(self.rect)
            width = int(x2 - x1)
            height = int(y2 - y1)
            if width < MIN_SIZE or height < MIN_SIZE:
                self.result = "CANCELLED"
            else:
                self.result = json.dumps(
                    {
                        "left": int(x1),
                        "top": int(y1),
                        "width": width,
                        "height": height,
                    }
                )
        self.root.destroy()


def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("--left", type=int, default=None)
    parser.add_argument("--top", type=int, default=None)
    parser.add_argument("--width", type=int, default=None)
    parser.add_argument("--height", type=int, default=None)
    return parser.parse_args()


def main():
    args = parse_args()
    initial_rect = None
    if (
        args.left is not None
        and args.top is not None
        and args.width is not None
        and args.height is not None
        and args.width > 0
        and args.height > 0
    ):
        initial_rect = [
            args.left,
            args.top,
            args.left + args.width,
            args.top + args.height,
        ]

    root = tk.Tk()
    root.title("OCR Region Picker")
    root.attributes("-fullscreen", True)
    root.attributes("-topmost", True)
    root.configure(bg="#000000")

    overlay = RegionOverlay(root, initial_rect)
    root.mainloop()
    print(overlay.result)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/ocr_region_overlay.swift
import AppKit
import Foundation

enum OutputResult: String {
  case cancelled = "CANCELLED"
  case fullscreen = "FULLSCREEN"
}

let SAFETY_TIMEOUT_SECONDS: TimeInterval = 30

enum DragMode {
  case none
  case create
  case move
  case resizeLeft
  case resizeRight
  case resizeTop
  case resizeBottom
  case resizeTopLeft
  case resizeTopRight
  case resizeBottomLeft
  case resizeBottomRight
}

final class OverlayWindow: NSWindow {
  override var canBecomeKey: Bool { true }
  override var canBecomeMain: Bool { true }
}

final class OverlayView: NSView {
  private let handleSize: CGFloat = 8
  private let minSize: CGFloat = 20
  private var selection: NSRect?
  private var baseSelection: NSRect?
  private var dragMode: DragMode = .none
  private var anchorPoint = NSPoint.zero
  private let onFinish: (String) -> Void
  private let screenFrame: NSRect
  private var safetyTimer: Timer?

  init(frame: NSRect, initial: NSRect?, onFinish: @escaping (String) -> Void) {
    self.screenFrame = frame
    self.onFinish = onFinish
    if let initial {
      self.selection = initial.standardized
    }
    super.init(frame: frame)
    wantsLayer = true
    scheduleSafetyTimeout()
  }

  required init?(coder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }

  override var acceptsFirstResponder: Bool { true }

  override func draw(_ dirtyRect: NSRect) {
    NSColor(calibratedWhite: 0, alpha: 0.45).setFill()
    bounds.fill()

    let help = "Arraste para criar area | arraste dentro para mover | arraste bordas/cantos para redimensionar | Enter confirma | F tela inteira | Esc cancela | timeout 30s"
    let attrs: [NSAttributedString.Key: Any] = [
      .foregroundColor: NSColor.white,
      .font: NSFont.systemFont(ofSize: 14, weight: .bold)
    ]
    help.draw(at: NSPoint(x: 20, y: bounds.height - 34), withAttributes: attrs)

    guard let rect = selection else { return }
    NSColor.clear.setFill()
    rect.fill()

    let stroke = NSBezierPath(rect: rect)
    NSColor.systemTeal.setStroke()
    stroke.lineWidth = 3
    stroke.stroke()

    let sizeText = "\(Int(rect.width))x\(Int(rect.height))"
    sizeText.draw(
      at: NSPoint(x: rect.minX, y: min(bounds.height - 52, rect.maxY + 8)),
      withAttributes: attrs
    )
    drawHandle(x: rect.minX, y: rect.minY)
    drawHandle(x: rect.maxX, y: rect.minY)
    drawHandle(x: rect.minX, y: rect.maxY)
    drawHandle(x: rect.maxX, y: rect.maxY)
  }

  private func drawHandle(x: CGFloat, y: CGFloat) {
    let handleRect = NSRect(
      x: x - handleSize,
      y: y - handleSize,
      width: handleSize * 2,
      height: handleSize * 2
    )
    NSColor.systemTeal.setFill()
    NSBezierPath(rect: handleRect).fill()
  }

  override func mouseDown(with event: NSEvent) {
    resetSafetyTimeout()
    if event.clickCount == 2 {
      confirmSelection()
      return
    }
    let point = convert(event.locationInWindow, from: nil)
    anchorPoint = point
    baseSelection = selection
    dragMode = hitTestMode(point: point) ?? .create
    if dragMode == .create {
      selection = NSRect(x: point.x, y: point.y, width: 1, height: 1)
    }
    needsDisplay = true
  }

  override func mouseDragged(with event: NSEvent) {
    resetSafetyTimeout()
    let point = clampPoint(convert(event.locationInWindow, from: nil))
    guard var current = selection else { return }

    switch dragMode {
    case .create:
      current = NSRect(
        x: min(anchorPoint.x, point.x),
        y: min(anchorPoint.y, point.y),
        width: abs(point.x - anchorPoint.x),
        height: abs(point.y - anchorPoint.y)
      )
    case .move:
      guard let base = baseSelection else { break }
      let dx = point.x - anchorPoint.x
      let dy = point.y - anchorPoint.y
      current = NSRect(x: base.origin.x + dx, y: base.origin.y + dy, width: base.width, height: base.height)
      current = clampRect(current)
    default:
      guard let base = baseSelection else { break }
      current = resizeRect(base: base, point: point, mode: dragMode)
    }

    selection = enforceMinSize(clampRect(current))
    needsDisplay = true
  }

  override func mouseUp(with event: NSEvent) {
    resetSafetyTimeout()
    _ = event
    dragMode = .none
    if let rect = selection {
      selection = enforceMinSize(clampRect(rect))
    }
    needsDisplay = true
  }

  override func rightMouseDown(with event: NSEvent) {
    resetSafetyTimeout()
    _ = event
    finish(with: OutputResult.cancelled.rawValue)
  }

  override func keyDown(with event: NSEvent) {
    if handleKeyEvent(event) {
      return
    }
  }

  @discardableResult
  func handleKeyEvent(_ event: NSEvent) -> Bool {
    resetSafetyTimeout()
    guard let chars = event.charactersIgnoringModifiers?.lowercased() else { return false }
    if event.keyCode == 53 { // esc
      finish(with: OutputResult.cancelled.rawValue)
      return true
    }
    if event.keyCode == 36 || event.keyCode == 76 { // enter
      confirmSelection()
      return true
    }
    if chars == "f" {
      finish(with: OutputResult.fullscreen.rawValue)
      return true
    }
    if chars == "q" {
      finish(with: OutputResult.cancelled.rawValue)
      return true
    }
    return false
  }

  private func scheduleSafetyTimeout() {
    safetyTimer?.invalidate()
    safetyTimer = Timer.scheduledTimer(withTimeInterval: SAFETY_TIMEOUT_SECONDS, repeats: false) { [weak self] _ in
      self?.finish(with: OutputResult.cancelled.rawValue)
    }
  }

  private func resetSafetyTimeout() {
    scheduleSafetyTimeout()
  }

  private func finish(with result: String) {
    safetyTimer?.invalidate()
    safetyTimer = nil
    onFinish(result)
  }

  private func confirmSelection() {
    guard let rect = selection, rect.width >= minSize, rect.height >= minSize else {
      finish(with: OutputResult.cancelled.rawValue)
      return
    }
    let standardized = rect.standardized
    let topFromTopLeft = Int(screenFrame.height - standardized.maxY)
    let json: [String: Int] = [
      "left": Int(standardized.minX),
      "top": topFromTopLeft,
      "width": Int(standardized.width),
      "height": Int(standardized.height),
      // Pontos da tela principal (para converter retângulo em pixels da captura Retina).
      "screenWidth": Int(screenFrame.width),
      "screenHeight": Int(screenFrame.height),
    ]
    guard
      let data = try? JSONSerialization.data(withJSONObject: json),
      let output = String(data: data, encoding: .utf8)
    else {
      finish(with: OutputResult.cancelled.rawValue)
      return
    }
    finish(with: output)
  }

  private func clampPoint(_ point: NSPoint) -> NSPoint {
    NSPoint(
      x: min(max(0, point.x), bounds.width),
      y: min(max(0, point.y), bounds.height)
    )
  }

  private func clampRect(_ rect: NSRect) -> NSRect {
    var r = rect
    if r.minX < 0 { r.origin.x = 0 }
    if r.minY < 0 { r.origin.y = 0 }
    if r.maxX > bounds.width { r.origin.x = bounds.width - r.width }
    if r.maxY > bounds.height { r.origin.y = bounds.height - r.height }
    r.origin.x = min(max(0, r.origin.x), bounds.width - r.width)
    r.origin.y = min(max(0, r.origin.y), bounds.height - r.height)
    return r
  }

  private func enforceMinSize(_ rect: NSRect) -> NSRect {
    var r = rect
    if r.width < minSize { r.size.width = minSize }
    if r.height < minSize { r.size.height = minSize }
    if r.maxX > bounds.width { r.origin.x = bounds.width - r.width }
    if r.maxY > bounds.height { r.origin.y = bounds.height - r.height }
    return r
  }

  private func hitTestMode(point: NSPoint) -> DragMode? {
    guard let rect = selection?.standardized else { return nil }
    let nearLeft = abs(point.x - rect.minX) <= handleSize
    let nearRight = abs(point.x - rect.maxX) <= handleSize
    let nearBottom = abs(point.y - rect.minY) <= handleSize
    let nearTop = abs(point.y - rect.maxY) <= handleSize
    let insideX = point.x >= rect.minX && point.x <= rect.maxX
    let insideY = point.y >= rect.minY && point.y <= rect.maxY

    if nearLeft && nearTop { return .resizeTopLeft }
    if nearRight && nearTop { return .resizeTopRight }
    if nearLeft && nearBottom { return .resizeBottomLeft }
    if nearRight && nearBottom { return .resizeBottomRight }
    if nearLeft && insideY { return .resizeLeft }
    if nearRight && insideY { return .resizeRight }
    if nearTop && insideX { return .resizeTop }
    if nearBottom && insideX { return .resizeBottom }
    if insideX && insideY { return .move }
    return nil
  }

  private func resizeRect(base: NSRect, point: NSPoint, mode: DragMode) -> NSRect {
    var left = base.minX
    var right = base.maxX
    var bottom = base.minY
    var top = base.maxY

    switch mode {
    case .resizeLeft, .resizeTopLeft, .resizeBottomLeft:
      left = point.x
    default:
      break
    }
    switch mode {
    case .resizeRight, .resizeTopRight, .resizeBottomRight:
      right = point.x
    default:
      break
    }
    switch mode {
    case .resizeBottom, .resizeBottomLeft, .resizeBottomRight:
      bottom = point.y
    default:
      break
    }
    switch mode {
    case .resizeTop, .resizeTopLeft, .resizeTopRight:
      top = point.y
    default:
      break
    }

    let normalizedLeft = min(left, right)
    let normalizedRight = max(left, right)
    let normalizedBottom = min(bottom, top)
    let normalizedTop = max(bottom, top)
    return NSRect(
      x: normalizedLeft,
      y: normalizedBottom,
      width: normalizedRight - normalizedLeft,
      height: normalizedTop - normalizedBottom
    )
  }
}

func parseArg(named name: String) -> Int? {
  guard let idx = CommandLine.arguments.firstIndex(of: name), idx + 1 < CommandLine.arguments.count else {
    return nil
  }
  return Int(CommandLine.arguments[idx + 1])
}

let app = NSApplication.shared
app.setActivationPolicy(.regular)
let screen = NSScreen.main ?? NSScreen.screens.first
guard let screen else {
  print(OutputResult.cancelled.rawValue)
  exit(0)
}
let frame = screen.frame

let initialRect: NSRect? = {
  guard
    let left = parseArg(named: "--left"),
    let top = parseArg(named: "--top"),
    let width = parseArg(named: "--width"),
    let height = parseArg(named: "--height"),
    width > 0,
    height > 0
  else { return nil }
  let yFromBottom = Int(frame.height) - top - height
  return NSRect(x: left, y: yFromBottom, width: width, height: height)
}()

let window = OverlayWindow(
  contentRect: frame,
  styleMask: [.borderless],
  backing: .buffered,
  defer: false
)
window.level = .screenSaver
window.collectionBehavior = [.canJoinAllSpaces, .fullScreenAuxiliary]
window.isOpaque = false
window.backgroundColor = .clear
window.ignoresMouseEvents = false
window.makeKeyAndOrderFront(nil)

let overlay = OverlayView(frame: frame, initial: initialRect) { result in
  print(result)
  NSApplication.shared.terminate(nil)
}

window.contentView = overlay
window.makeFirstResponder(overlay)
app.activate(ignoringOtherApps: true)
window.makeKeyAndOrderFront(nil)
_ = NSEvent.addLocalMonitorForEvents(matching: .keyDown) { event in
  if overlay.handleKeyEvent(event) {
    return nil
  }
  return event
}
app.run()
